buildtest/buildsystem/base.py

```diff
@@ -0,0 +1,558 @@
+        """
+        self.recipe = None
+
+        # Read the lookup to get schemas available
+        self.lookup = get_schemas_available()
+
+        # Load the configuration file, fails on any error
+        self.load(config_file)
+
+    def load(self, config_file):
+        """load a config file. We check that it exists, and that it is valid.
+           we exit on any error, as the config is not loadable
+
+           Parameters
+           ==========
+           config_file: the pull path to the configuration file, must exist.
+        """
+        self.config_file = os.path.abspath(config_file)
+
+        if not os.path.exists(self.config_file):
+            sys.exit("Build configuration %s does not exist." % self.config_file)
+
+        elif os.path.isdir(self.config_file):
+            sys.exit(
+                "Please provide a file path (not a directory path) to a build recipe."
+            )
+
+        # all test configs must pass global validation (sets self.recipe)
+        self._validate_global()
+
+        # validate each schema defined in the recipes
+        self._validate()
+
+    def __str__(self):
+        return "[buildtest-build-config]"
+
+    def __repr__(self):
+        return "[buildtest-build-config]"
+
+    # Validation
+
+    def _validate(self):
+        """Given a loaded recipe, validate that the type is known in the lookup
+           to buildtest. If a version is provided, honor it. If not, use latest.
+           We also don't allow repeated keys in the same file.
+        """
+        version = self.recipe.get("version", "latest")
+        seen = set()
+        for name, section in self.recipe.items():
+
+            if name == "version":
+                continue
+
+            # Check for repeated keys
+            elif name in seen:
+                sys.exit("Invalid recipe: %s is repeated more than once." % name)
+            seen.add(name)
+
+            # Ensure we have a recipe for it
+            if section["type"] not in self.lookup.keys():
+                sys.exit("type %s is not known to buildtest." % section["type"])
+
+            # And that there is a version file
+            if version not in self.lookup[section["type"]]:
+                sys.exit(
+                    "version %s is not known for type %s. Try using latest." % version,
+                    section["type"],
+                )
+
+            # Finally, validate the section against the schema
+            schema_file = os.path.join(
+                here, section["type"], self.lookup[section["type"]][version]
+            )
+            validate(instance=section, schema=load_schema(schema_file))
+
+    def _validate_global(self, config_file=None):
+        """the global validation ensures that the overall structure of the
+           file is sound for further parsing. We load in the outer.schema.json
+           for this purpose. The function also allows a custom config to be 
+           to extend the usage of the class.
```

Comment:
can you define the parameter ``config_files`` in docstring

Reply:
Let's handle all docstrings in another PR, there is too much linting / fixes that need to be done and since it's globally across the library it's scoped outside of this work.